loader/wsi.c

```diff
@@ -600,6 +794,54 @@ VkResult allocate_icd_surface_struct(struct loader_instance *instance, size_t ba
     return res;
 }
 
+VkResult copy_surface_create_info(struct loader_instance *loader_inst, VkIcdSurface *icd_surface, const void *create_info,
+                                  size_t struct_type_info_count, const struct loader_struct_type_info *struct_type_info,
+                                  const char *base_struct_name) {
+    size_t create_info_total_size = 0;
+    const VkBaseInStructure *base = (const VkBaseInStructure *)create_info;
+    while (NULL != base) {
+        bool known_struct = false;
+        for (size_t i = 0; i < struct_type_info_count; ++i) {
+            if (base->sType == struct_type_info[i].stype) {
+                create_info_total_size += loader_aligned_size(struct_type_info[i].size);
+                base = base->pNext;
+                known_struct = true;
+                break;
+            }
+        }
```

Comment:
Reinterpreting a type through a void* is UB from my understanding, and the way around it is to use memcpy. Since here we don't modify the chain, it is a simple fix.
```suggestion
    const void *pNext = create_info;
    while (NULL != pNext) {
        VkBaseInStructure base = {0};
        memcpy(&base, pNext, sizeof(VkBaseInStructure));
        bool known_struct = false;
        for (size_t i = 0; i < struct_type_info_count; ++i) {
            if (base.sType == struct_type_info[i].stype) {
                create_info_total_size += loader_aligned_size(struct_type_info[i].size);
                pNext = base.pNext;
                known_struct = true;
                break;
            }
        }
```


That said, because the pointer really points to a union, its a more tricky situation. Normally type punning through union members is not valid so the memcpy is needed, but maybe it isn't? This is rather complex and may not be necessary when code-gen is used (because then it isn't using arrays of uint8's but concrete types that the compiler understands.

Reply:
> Reinterpreting a type through a void* is UB from my understanding, and the way around it is to use memcpy. Since here we don't modify the chain, it is a simple fix.

Well, not exactly. The primary reason for UB in these cases is due to alignment differences based on the base type, which is why I align the structs as is.

If we'd be strict, nothing except reinterpret casting between any pointer type and char pointers is UB (casting to/from char pointers is not UB exactly because chars don't require higher than single-byte-alignment).

But sure, I can change this to memcpy, we could do better than this with codegen in the future anyway.

Reply:
I think it is fine to leave as is - these endless "is it UB or is not not UB" gets tiring and doesn't always yield actionable results. Difficult to diagnose crashes make me hesitant around UB, but I think the aligned structs should obviate most of those issues.